src/ape/api/query.py

```diff
@@ -0,0 +1,79 @@
+from typing import Any, Dict, List, Literal, Optional, Union
+
+import pandas as pd
+from pydantic import BaseModel
+
+from ape.types import AddressType
+from ape.utils import abstractdataclass, abstractmethod
+
+Query = Union["BlockQuery", "AccountQuery", "ContractEventQuery", "ContractMethodQuery"]
+
+
+class BaseQuery(BaseModel):
+    type: str  # Used as discriminator
+    columns: List[str]
+    filter_args: Dict[str, Any]
+    engine_to_use: Optional[str] = None
+
+
+class BaseBlockQuery(BaseQuery):
+    start_block: int = 0
+    stop_block: Optional[int] = None
+
+
+class BlockQuery(BaseBlockQuery):
+    type: Literal["blocks"]
+
+
+class BaseAccountQuery(BaseModel):
+    start_nonce: int = 0
+    stop_nonce: Optional[int] = None
+    engine_to_use: Optional[str] = None
+
+
+class AccountQuery(BaseBlockQuery):
+    type: Literal["accounts"]
+    account: AddressType
+
+
+class ContractEventQuery(BaseBlockQuery):
+    type: Literal["contract_events"]
+    contract: AddressType
+    event_id: bytes
+
+
+class ContractMethodQuery(BaseBlockQuery):
+    type: Literal["contract_calls"]
+    contract: AddressType
+    method_id: bytes
+    method_args: Dict[str, Any]
+
+
+@abstractdataclass
+class QueryAPI:
+
+    engine_to_use: Optional[str] = None
```

Comment:
@unparalleled-js brought up a good suggestions to add this as a way of overriding the query_engine choice provided by performance. If you already know you want to use postgres this should allow you to short-circuit the logic to access it. I'm very open to renaming things as well, but figured this is just a first pass to get some basic structure in place.

Reply:
I like this, just see comments I made to better integrate this feature